File: skeleton/src/main/java/me/shkschneider/skeleton/ui/ImageHelper.java
package me.shkschneider.skeleton.ui;

import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Shader;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.util.DisplayMetrics;
import android.view.View;
import android.widget.RelativeLayout;

import me.shkschneider.skeleton.helpers.ApplicationHelper;
import me.shkschneider.skeleton.helpers.LogHelper;

public class ImageHelper {

    @Deprecated
    public static Drawable fromBitmap(@NonNull final Bitmap bitmap) {
        return new BitmapDrawable(bitmap);
    }

    public static Bitmap fromDrawable(@NonNull final Drawable drawable) {
        if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable) drawable).getBitmap();
        }

        final Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return bitmap;
    }

    public static Bitmap circular(@NonNull final Bitmap bitmap) {
        final Bitmap circleBitmap = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), Bitmap.Config.ARGB_8888);
        final BitmapShader shader = new BitmapShader (bitmap, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP);
        final Paint paint = new Paint();
        paint.setShader(shader);
        final Canvas canvas = new Canvas(circleBitmap);
        canvas.drawCircle(bitmap.getWidth() / 2, bitmap.getHeight() / 2, bitmap.getWidth() / 2, paint);
        return circleBitmap;
    }

    public static Bitmap rotate(@NonNull final Bitmap bitmap, final float degrees) {
        final Matrix matrix = new Matrix();
        matrix.postRotate(degrees);
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.getWidth(), bitmap.getHeight(), matrix, true);
    }

    public static Bitmap fromView(@NonNull final View view) {
        final DisplayMetrics displayMetrics = ApplicationHelper.resources().getDisplayMetrics();
        if (displayMetrics == null) {
            LogHelper.warning("DisplayMetrics was NULL");
            return null;
        }

        view.setLayoutParams(new RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT));
        view.measure(displayMetrics.widthPixels, displayMetrics.heightPixels);
        view.layout(0, 0, displayMetrics.widthPixels, displayMetrics.heightPixels);
        view.buildDrawingCache();

        final Bitmap bitmap = Bitmap.createBitmap(view.getMeasuredWidth(), view.getMeasuredHeight(), Bitmap.Config.ARGB_8888);
        final Canvas canvas = new Canvas(bitmap);
        view.draw(canvas);
        return bitmap;
    }

    // TODO palette
//    public static int vibrantColor(@NonNull final Bitmap bitmap) {
//        final Palette palette = Palette.generate(bitmap);
//        return palette.getVibrantColor(ApplicationHelper.resources().getColor(R.color.transparent));
//    }

    // TODO palette
//    public static int mutedColor(@NonNull final Bitmap bitmap) {
//        final Palette palette = Palette.generate(bitmap);
//        return palette.getMutedColor(ApplicationHelper.resources().getColor(R.color.transparent));
//    }

}

File: library/src/main/java/me/shkschneider/skeleton/ui/ImageHelper.kt
package me.shkschneider.skeleton.ui

import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.drawable.BitmapDrawable
import android.os.Build
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.RoundedBitmapDrawableFactory
import android.view.View
import android.widget.RelativeLayout
import androidx.annotation.DrawableRes
import me.shkschneider.skeleton.helper.ApplicationHelper
import me.shkschneider.skeleton.helper.ContextHelper
import me.shkschneider.skeleton.helperx.Metrics

object ImageHelper {

    object Bitmap {

        fun fromResource(@DrawableRes id: Int): android.graphics.Bitmap? {
            return BitmapFactory.decodeResource(ApplicationHelper.resources(), id)
        }

        fun fromDrawable(drawable: android.graphics.drawable.Drawable): android.graphics.Bitmap {
            if (drawable is BitmapDrawable) {
                return drawable.bitmap
            }
            val bitmap = android.graphics.Bitmap.createBitmap(drawable.intrinsicWidth, drawable.intrinsicHeight, android.graphics.Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            drawable.setBounds(0, 0, canvas.width, canvas.height)
            drawable.draw(canvas)
            return bitmap
        }

        fun fromView(view: View): android.graphics.Bitmap? {
            val displayMetrics = Metrics.displayMetrics()
            view.layoutParams = RelativeLayout.LayoutParams(RelativeLayout.LayoutParams.WRAP_CONTENT, RelativeLayout.LayoutParams.WRAP_CONTENT)
            view.measure(displayMetrics.widthPixels, displayMetrics.heightPixels)
            view.layout(0, 0, displayMetrics.widthPixels, displayMetrics.heightPixels)
            if (Build.VERSION.SDK_INT < 28) {
                /**
                 * This method was deprecated in API level 28. The view drawing cache was largely
                 * made obsolete with the introduction of hardware-accelerated rendering in API 11.
                 */
                @Suppress("DEPRECATION")
                view.buildDrawingCache()
            }
            val bitmap = android.graphics.Bitmap.createBitmap(view.measuredWidth, view.measuredHeight, android.graphics.Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            view.draw(canvas)
            return bitmap
        }

    }

    object Drawable {

        fun fromResource(@DrawableRes id: Int): android.graphics.drawable.Drawable? {
            return ContextCompat.getDrawable(ContextHelper.applicationContext(), id)
        }

        fun fromBitmap(bitmap: android.graphics.Bitmap) : android.graphics.drawable.Drawable? {
            return BitmapDrawable(ApplicationHelper.resources(), bitmap)
        }

    }

    fun circular(bitmap: android.graphics.Bitmap) : android.graphics.drawable.Drawable {
        val roundedBitmapDrawable = RoundedBitmapDrawableFactory.create(ApplicationHelper.resources(), bitmap)
        roundedBitmapDrawable.isCircular = true
        roundedBitmapDrawable.setAntiAlias(true)
        return roundedBitmapDrawable
    }

}
